src/tribler/core/upgrade/tribler_db/migration_chain.py

```diff
@@ -0,0 +1,53 @@
+import logging
+from typing import Callable, Iterator, List, Optional
+
+from tribler.core.components.database.db.tribler_database import TriblerDatabase
+from tribler.core.upgrade.tribler_db.decorator import has_migration_metadata
+from tribler.core.upgrade.tribler_db.scheme_migrations.scheme_migration_0 import scheme_migration_0
+from tribler.core.utilities.path_util import Path
+from tribler.core.utilities.simpledefs import STATEDIR_DB_DIR
+
+
+class TriblerDatabaseMigrationChain:
+    """ A chain of migrations that can be executed on a TriblerDatabase.
+
+    To create a new migration, create a new function and decorate it with the `migration` decorator. Then add it to
+    the `DEFAULT_CHAIN` list.
+    """
+
+    DEFAULT_CHAIN = [
+        scheme_migration_0,
+        # add your migration here
+    ]
+
+    def __init__(self, state_dir: Path, chain: Optional[List[Callable]] = None):
+        self.logger = logging.getLogger(self.__class__.__name__)
+        self.state_dir = state_dir
+
+        db_path = self.state_dir / STATEDIR_DB_DIR / 'tribler.db'
+        self.logger.info(f'Tribler DB path: {db_path}')
+        self.db = TriblerDatabase(str(db_path), check_tables=False) if db_path.is_file() else None
+
+        self.migrations = chain or self.DEFAULT_CHAIN
+
+    def execute(self) -> bool:
+        """ Execute all migrations in the chain.
+
+        Returns: True if all migrations were executed successfully, False otherwise.
+        An exception in any of the migrations will halt the execution chain and be re-raised.
+        """
+
+        if not self.db:
+            return False
+
+        for _ in self.steps():
+            ...
```

Comment:
Is it intended to be this way in the current PR?

Reply:
Yes, this executes each migration step-by-step. Do you have a better approach to achieve the same?

Reply:
Currently, the migration is executed as a side effect of the iteration. It looks non-intuitive. I think it is better to split it into two different phases: the selection of suitable migrations and the execution of selected migrations:
```python
for migration in self.find_migrations_to_apply(current_version=self.db.version):
    migration(self.db, state_dir=self.state_dir)
```

This way, it becomes possible to filter out migrations that were already applied and only apply new migrations. It also allows to separately test the selection of suitable migrations and the execution of a migration.

Reply:
Indeed it is just another approach to implementing the same feature. The current version performs the "selection" procedure inside the decorator. It is already tested in `test_decorator.py`

> This way, it becomes possible to filter out migrations that were already applied and only apply new migrations. It also allows to separately test the selection of suitable migrations and the execution of a migration.

The current approach already contains these properties.

Reply:
> Currently, the migration is executed as a side effect of the iteration

That was the intention. This way we can test the execution step-by-step, just iterating the steps.

But I'm also with rewriting it to something like:

```python
    def execute(self) -> bool:
        """ Execute all migrations in the chain.

        Returns: True if all migrations were executed successfully, False otherwise.
        An exception in any of the migrations will halt the execution chain and be re-raised.
        """

        if not self.db:
            return False

        for m in self.select():
            m(self.db, state_dir=self.state_dir)

        return True

    def select(self) -> Iterator:
        for m in self.migrations:
            if not has_migration_metadata(m):
                raise NotImplementedError(f'The migration {m} should have `migration` decorator')
            yield m
```

Reply:
I've removed the `steps` property. It's simpler and more straightforward now, but we've lost the ability to execute steps one by one and check the results of each execution.

The last version:

```pyhon
    def execute(self) -> bool:
        """ Execute all migrations in the chain.

        Returns: True if all migrations were executed successfully, False otherwise.
        An exception in any of the migrations will halt the execution chain and be re-raised.
        """

        if not self.db:
            return False

        for m in self.migrations:
            if not has_migration_metadata(m):
                raise NotImplementedError(f'The migration {m} should have `migration` decorator')
            m(self.db, state_dir=self.state_dir)

        return True
```